processor/src/main/java/com/linecorp/decaton/processor/runtime/internal/CompletionImpl.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor.runtime.internal;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+import com.linecorp.decaton.processor.runtime.Completion;
+
+import lombok.Data;
+import lombok.experimental.Accessors;
+
+@Accessors(fluent = true)
+@Data
+public class CompletionImpl implements Completion {
+    private final CompletableFuture<Void> future;
+    private Supplier<TimeoutChoice> expireCallback;
+    private volatile Completion dependency;
+
+    public static CompletionImpl completedCompletion() {
+        CompletionImpl comp = new CompletionImpl();
+        comp.complete();
+        return comp;
+    }
+
+    public CompletionImpl() {
+        future = new CompletableFuture<>();
+    }
+
+    @Override
+    public boolean isComplete() {
+        return future.isDone();
+    }
+
+    @Override
+    public CompletionStage<Void> asFuture() {
+        return future;
+    }
+
+    @Override
+    public boolean tryExpire() {
+        if (isComplete()) {
+            return true;
+        }
+        if (dependency != null) {
+            if (!dependency.tryExpire()) {
+                return false;
+            }
+            dependency = null;
+        }
+        if (expireCallback.get() == TimeoutChoice.GIVE_UP) {
+            complete();
```

Comment:
Say we have processors like below:
```
- CompactionProcessor: deferCompletion(() -> GIVE_UP).completeWith(ctx.push())
  - UserProcessor1: deferCompletion(() -> EXTEND).completeWith(ctx.push())
    - UserPostProcessor2: deferCompletion(() -> GIVE_UP).completeWith(doSomethingOnAsyncFramework())
```

In current impl, since "leaf" processor's TimeoutChoice is GIVE_UP, expiration isn't extended when timeout happens, despite UserProcessor1 sets TimeoutChoice to EXTEND.

Is that intentional?

Reply:
No... In that case, after 2 is expired 1's callback should be called, hence it should extend timeout. Let me fix that, super nice catch.